zipkin-server/src/main/java/zipkin2/server/internal/throttle/ThrottledCall.java

```diff
@@ -37,183 +38,181 @@
-      limitListener.onIgnore();
-      throw new RuntimeException("Interrupted while blocking on a throttled call", e);
-    } catch (RuntimeException | Error e) {
-      propagateIfFatal(e);
-      // Ignoring in all cases here because storage itself isn't saying we need to throttle.  Though, we may still be
-      // write bound, but a drop in concurrency won't necessarily help.
-      limitListener.onIgnore();
-      throw e;
+  /**
+   * To simplify code, this doesn't actually invoke the underlying {@link #execute()} method. This
+   * is ok because in almost all cases, doing so would imply invoking {@link #enqueue(Callback)}
+   * anyway.
+   */
+  @Override protected Void doExecute() throws IOException {
+    AwaitableCallback awaitableCallback = new AwaitableCallback();
+    doEnqueue(awaitableCallback);
+    if (!await(awaitableCallback.countDown)) throw new InterruptedIOException();
+
+    Throwable t = awaitableCallback.throwable.get();
+    if (t != null) {
+      if (t instanceof Error) throw (Error) t;
+      if (t instanceof IOException) throw (IOException) t;
+      if (t instanceof RuntimeException) throw (RuntimeException) t;
+      throw new RuntimeException(t);
     }
+    return null;
   }
 
-  @Override protected void doEnqueue(Callback<V> callback) {
-    Listener limitListener = limiter.acquire(null)
+  @Override protected void doEnqueue(Callback<Void> callback) {
+    Listener limiterListener = limiter.acquire(null)
       .orElseThrow(RejectedExecutionException::new); // TODO: make an exception message
+    LimiterReleasingCallback releasingCallback =
+      new LimiterReleasingCallback(callback, isOverCapacity, limiterListener);
 
     try {
-      executor.execute(new QueuedCall<>(this, callback, limitListener));
-    } catch (RuntimeException | Error e) {
-      propagateIfFatal(e);
-      // Ignoring in all cases here because storage itself isn't saying we need to throttle.  Though, we may still be
-      // write bound, but a drop in concurrency won't necessarily help.
-      limitListener.onIgnore();
-      throw e;
+      executor.execute(new EnqueueAndAwait(this, releasingCallback));
+    } catch (RuntimeException | Error t) { // possibly rejected, but from the executor, not storage!
+      propagateIfFatal(t);
+      callback.onError(t);
+      // Ignoring in all cases here because storage itself isn't saying we need to throttle. Though
+      // we may still be write bound, but a drop in concurrency won't necessarily help.
+      limiterListener.onIgnore();
+      throw t;
     }
   }
 
-  @Override public Call<V> clone() {
-    return new ThrottledCall<>(executor, limiter, isOverCapacity, delegate.clone());
+  @Override public Call<Void> clone() {
+    return new ThrottledCall(delegate.clone(), executor, limiter, isOverCapacity);
   }
 
   @Override public String toString() {
     return "Throttled(" + delegate + ")";
   }
 
-  static String setCurrentThreadName(String name) {
```

Comment:
I removed this (eventhough I asked for it) because actually it didn't help at all and so was a bit distracting, and also untested. I want to focus lines on tested code that helps troubleshooting